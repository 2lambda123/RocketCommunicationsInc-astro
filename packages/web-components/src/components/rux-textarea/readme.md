# rux-textarea

<!-- Auto Generated Below -->


## Properties

| Property      | Attribute     | Description                                                                                                                                                                             | Type                                          | Default     |
| ------------- | ------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------- | ----------- |
| `disabled`    | `disabled`    | Disables the button via HTML disabled attribute. Button takes on a distinct visual state. Cursor uses the not-allowed system replacement and all keyboard and mouse events are ignored. | `boolean`                                     | `false`     |
| `errorText`   | `error-text`  | The validation error text                                                                                                                                                               | `string \| undefined`                         | `undefined` |
| `helpText`    | `help-text`   | The  or explanation text                                                                                                                                                                | `string \| undefined`                         | `undefined` |
| `invalid`     | `invalid`     | Presentational only. Renders the Textarea as invalid.                                                                                                                                   | `boolean`                                     | `false`     |
| `label`       | `label`       | The textarea label text. For HTML content, use the `label` slot instead.                                                                                                                | `string \| undefined`                         | `undefined` |
| `maxLength`   | `max-length`  | The input maxLength attribute                                                                                                                                                           | `string \| undefined`                         | `undefined` |
| `minLength`   | `min-length`  | The input minLength attribute                                                                                                                                                           | `string \| undefined`                         | `undefined` |
| `name`        | `name`        | The input name                                                                                                                                                                          | `string`                                      | `''`        |
| `placeholder` | `placeholder` | The textarea placeholder text                                                                                                                                                           | `string \| undefined`                         | `undefined` |
| `required`    | `required`    | Sets the input as disabled                                                                                                                                                              | `boolean`                                     | `false`     |
| `rows`        | `rows`        | The input rows attribute                                                                                                                                                                | `number \| undefined`                         | `undefined` |
| `size`        | `size`        | Styles the input element size between small, medium and large. The default styling is medium.                                                                                           | `"large" \| "medium" \| "small" \| undefined` | `undefined` |
| `value`       | `value`       | The input value                                                                                                                                                                         | `string`                                      | `''`        |


## Events

| Event       | Description                                                                                                                                                                    | Type               |
| ----------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------------------ |
| `ruxblur`   | Fired when an element has lost focus - [HTMLElement/blur_event](https://developer.mozilla.org/en-US/docs/Web/API/Element/blur_event)                                           | `CustomEvent<any>` |
| `ruxchange` | Fired when the value of the input changes - [HTMLElement/input_event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event)                                | `CustomEvent<any>` |
| `ruxinput`  | Fired when an alteration to the input's value is committed by the user - [HTMLElement/change_event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event) | `CustomEvent<any>` |


## Slots

| Slot      | Description        |
| --------- | ------------------ |
| `"label"` | The textarea label |


## Shadow Parts

| Part           | Description                              |
| -------------- | ---------------------------------------- |
| `"error-text"` | The error text element                   |
| `"form-field"` | The form-field wrapper container         |
| `"help-text"`  | The help text element                    |
| `"label"`      | The input label when `label` prop is set |
| `"required"`   | The asterisk when required is true       |
| `"textarea"`   | The textarea element                     |


## CSS Custom Properties

| Name                                    | Description                                             |
| --------------------------------------- | ------------------------------------------------------- |
| `--textarea-background-color`           | Background color                                        |
| `--textarea-border-color`               | [DEPRECATED] Border color of the textarea.              |
| `--textarea-focus-border-color`         | [DEPRECATED] Border color while in a focused state      |
| `--textarea-invalid-border-color`       | Border color while in an invalid state                  |
| `--textarea-selection-background-color` | [DEPRECATED] Background color while in a selected state |
| `--textarea-text-color`                 | [DEPRECATED] Text color                                 |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
