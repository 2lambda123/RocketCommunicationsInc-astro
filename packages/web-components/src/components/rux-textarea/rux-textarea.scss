@use '../../common/functional-components/FormFieldMessage/form-field-message.scss';

:host {
    /**
     * @prop --textarea-border-color: [DEPRECATED] Border color of the textarea.
    */
    --textarea-border-color: var(--color-border-interactive-muted);

    /**
     * @prop --textarea-text-color: [DEPRECATED] Text color
    */
    --textarea-text-color: var(--color-text-primary);

    /** @prop --textarea-background-color: Background color */
    --textarea-background-color: var(--color-background-base-default);

    /** @prop --textarea-invalid-border-color: Border color while in an invalid state */
    --textarea-invalid-border-color: var(--status-symbol-color-fill-critical);

    /**
     * @prop --textarea-focus-border-color: [DEPRECATED] Border color while in a focused state
    */
    --textarea-focus-border-color: var(--color-background-interactive-hover);

    /**
     * @prop --textarea-selection-background-color: [DEPRECATED] Background color while in a selected state
    */
    --textarea-selection-background-color: var(--color-palette-brightblue-400);

    display: block;
}

.hidden,
:host([hidden]) {
    display: none;
}

.rux-textarea {
    box-sizing: border-box;
    -webkit-order: 2;
    order: 2;
    min-height: 4.25rem;
    width: 100%;
    padding: 0.469rem 0.5rem; //7.5px 8px
    border: 1px solid var(--color-border-interactive-muted);
    border-radius: var(--radius-base);
    font-family: var(--font-body-1-font-family);
    font-size: var(--font-body-1-font-size);
    letter-spacing: var(--font-body-1-letter-spacing);
    font-weight: var(--font-body-1-font-weight);
    color: var(--color-text-primary);
    background-color: var(--color-background-base-default);
    &--invalid {
        border: 1px solid var(--color-text-error);
    }
    &--disabled {
        opacity: 0.4;
        opacity: var(--opacity-disabled, 40%);
        cursor: not-allowed;
    }
    &--small {
        padding: 0.219rem 0.5rem; //3.5px 8px
        min-height: 3.25rem; //52px
    }
    &--large {
        min-height: 5.75rem; //92px
        padding: 0.844rem 0.5rem; //13.5px 8px
    }

    &:focus {
        border-color: var(--color-background-interactive-hover);
        outline: none;
    }
    &:hover {
        border-color: var(--color-background-interactive-hover);
        outline: none;
    }
    &--disabled:hover {
        border: 1px solid var(--color-border-interactive-muted);
        color: var(--color-text-primary);
    }
    &--invalid:hover {
        border: 1px solid var(--color-text-error);
    }
    &::placeholder {
        color: var(--color-text-placeholder);
    }
}

.rux-textarea-field {
    display: flex;
    flex-direction: column;
    color: var(--color-text-primary);
}

.rux-textarea-label {
    font-family: var(--font-body-1-font-family);
    font-size: var(--font-body-1-font-size);
    font-weight: var(--font-body-1-font-weight);
    letter-spacing: var(--font-body-1-letter-spacing);
    margin-bottom: var(--spacing-input-label-top);
    &__asterisk {
        margin-left: 4px;
    }
}

::selection {
    background-color: var(--color-palette-brightblue-400);
}
