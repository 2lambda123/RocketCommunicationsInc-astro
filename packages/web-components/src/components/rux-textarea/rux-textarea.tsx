import {
    Component,
    h,
    Event,
    EventEmitter,
    Host,
    Prop,
    Element,
    Watch,
    State,
    Method,
} from '@stencil/core'
import { FormFieldInterface } from '../../common/interfaces.module'
import { hasSlot, renderHiddenInput } from '../../utils/utils'

let id = 0

/**
 * @slot label - The textarea label
 * @slot help-text -  the help text
 * @slot error-text -  the error text
 * @part error-text - The error text element
 * @part form-field - The form-field wrapper container
 * @part help-text - The help text element
 * @part label - The input label when `label` prop is set
 * @part textarea - The textarea element
 * @part required - The asterisk when required is true
 */
@Component({
    tag: 'rux-textarea',
    styleUrl: 'rux-textarea.scss',
    shadow: true,
})
export class RuxTextarea implements FormFieldInterface {
    private inputId = `rux-textarea-${++id}`
    private textareaEl!: HTMLTextAreaElement
    @State() hasLabelSlot = false
    @State() hasHelpSlot = false
    @State() hasErrorSlot = false

    /**
     * The textarea label text. For HTML content, use the `label` slot instead.
     */
    @Prop() label?: string
    /**
     * The textarea placeholder text
     */
    @Prop() placeholder?: string

    /**
     * The  or explanation text
     */
    @Prop({ attribute: 'help-text' }) helpText?: string

    /**
     * The validation error text
     */
    @Prop({ attribute: 'error-text' }) errorText?: string

    /**
     * Presentational only. Renders the Textarea as invalid.
     */
    @Prop() invalid = false

    /**
     * The input value
     */
    @Prop({ mutable: true, reflect: true }) value: string = ''

    /**
     * The input name
     */
    @Prop() name = ''

    /**
     * The input minLength attribute
     */
    @Prop({ attribute: 'min-length' }) minLength?: string

    /**
     * The input maxLength attribute
     */
    @Prop({ attribute: 'max-length' }) maxLength?: string

    /**
     * The input rows attribute
     */
    @Prop() rows?: number

    /**
     * Disables the button via HTML disabled attribute. Button takes on a distinct visual state. Cursor uses the not-allowed system replacement and all keyboard and mouse events are ignored.
     */
    @Prop({ reflect: true }) disabled = false

    /**
     * Sets the input as disabled
     */
    @Prop() required: boolean = false

    /**
     * Styles the input element size between small, medium and large. The default styling is medium.
     */
    @Prop({ reflect: true }) size?: 'small' | 'medium' | 'large'

    /**
     * Fired when the value of the input changes - [HTMLElement/input_event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event)
     */
    @Event({ eventName: 'ruxchange' }) ruxChange!: EventEmitter

    /**
     * Fired when an alteration to the input's value is committed by the user - [HTMLElement/change_event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event)
     */
    @Event({ eventName: 'ruxinput' }) ruxInput!: EventEmitter
    /**
     * Fired when an element has lost focus - [HTMLElement/blur_event](https://developer.mozilla.org/en-US/docs/Web/API/Element/blur_event)
     */
    @Event({ eventName: 'ruxblur' }) ruxBlur!: EventEmitter

    @Element() el!: HTMLRuxTextareaElement

    /**
     * Sets element as focused
     */
    @Method()
    async setFocus(options?: FocusOptions) {
        this.textareaEl.focus(options)
    }

    @Watch('label')
    handleLabelChange() {
        this._handleSlotChange()
    }

    connectedCallback() {
        this._onChange = this._onChange.bind(this)
        this._onInput = this._onInput.bind(this)
        this._handleSlotChange = this._handleSlotChange.bind(this)
    }

    disconnectedCallback() {
        this.el!.shadowRoot!.removeEventListener(
            'slotchange',
            this._handleSlotChange
        )
    }

    componentWillLoad() {
        this._handleSlotChange()
    }

    get hasLabel() {
        return this.label ? true : this.hasLabelSlot
    }

    private _handleSlotChange() {
        this.hasLabelSlot = hasSlot(this.el, 'label')
        this.hasErrorSlot = hasSlot(this.el, 'error-text')
        this.hasHelpSlot = hasSlot(this.el, 'help-text')
    }

    private _onChange(e: Event) {
        const target = e.target as HTMLInputElement
        this.value = target.value
        this.ruxChange.emit()
    }

    private _onInput(e: Event) {
        const target = e.target as HTMLInputElement
        this.value = target.value
        this.ruxInput.emit()
    }

    private _onBlur = () => {
        this.ruxBlur.emit()
    }

    render() {
        renderHiddenInput(true, this.el, this.name, this.value, this.disabled)
        return (
            <Host>
                <div class="rux-textarea-field" part="form-field">
                    {this.hasLabel ? (
                        <label
                            class={{
                                'rux-textarea-label': true,
                            }}
                            aria-hidden={this.hasLabel ? 'false' : 'true'}
                            htmlFor={this.inputId}
                            part="label"
                        >
                            <span class={{ hidden: !this.hasLabel }}>
                                <slot
                                    onSlotchange={this._handleSlotChange}
                                    name="label"
                                >
                                    {this.label}
                                    {this.required && (
                                        <span
                                            part="required"
                                            class="rux-textarea-label__asterisk"
                                        >
                                            &#42;
                                        </span>
                                    )}
                                </slot>
                            </span>
                        </label>
                    ) : null}
                    <textarea
                        ref={(el) => (this.textareaEl = el!)}
                        name={this.name}
                        disabled={this.disabled}
                        aria-invalid={this.invalid ? 'true' : 'false'}
                        placeholder={this.placeholder}
                        required={this.required}
                        minlength={this.minLength}
                        maxlength={this.maxLength}
                        value={this.value}
                        class={{
                            'rux-textarea': true,
                            'rux-textarea--disabled': this.disabled,
                            'rux-textarea--invalid': this.invalid,
                            'rux-textarea--small': this.size === 'small',
                            'rux-textarea--large': this.size === 'large',
                        }}
                        id={this.inputId}
                        rows={this.rows}
                        onChange={this._onChange}
                        onInput={this._onInput}
                        onBlur={this._onBlur}
                        part="textarea"
                    ></textarea>
                </div>
                <div
                    class={{
                        'rux-error-text': !!this.errorText || this.hasErrorSlot,
                        hidden: !this.errorText && !this.hasErrorSlot,
                    }}
                    part="error-text"
                >
                    <svg
                        fill="none"
                        width="14"
                        height="14"
                        xmlns="http://www.w3.org/2000/svg"
                        viewBox="0 0 14 14"
                    >
                        <path
                            fill-rule="evenodd"
                            clip-rule="evenodd"
                            d="M11.393 12.25c.898 0 1.458-.974 1.009-1.75L8.009 2.91a1.166 1.166 0 0 0-2.018 0L1.598 10.5c-.449.776.111 1.75 1.01 1.75h8.784ZM7 8.167a.585.585 0 0 1-.583-.584V6.417c0-.321.262-.584.583-.584.32 0 .583.263.583.584v1.166c0 .321-.262.584-.583.584Zm-.583 1.166V10.5h1.166V9.333H6.417Z"
                            fill="currentColor"
                        />
                    </svg>
                    <slot
                        name="error-text"
                        onSlotchange={this._handleSlotChange}
                    >
                        {this.errorText}
                    </slot>
                </div>
                <div
                    class={{
                        'rux-help-text':
                            (!!this.helpText || this.hasHelpSlot) &&
                            (!this.errorText || !this.hasErrorSlot),
                        hidden:
                            (!this.helpText && !this.hasHelpSlot) ||
                            !!this.errorText ||
                            this.hasErrorSlot,
                    }}
                    part="help-text"
                >
                    <slot
                        name="help-text"
                        onSlotchange={this._handleSlotChange}
                    >
                        {this.helpText}
                    </slot>
                </div>
            </Host>
        )
    }
}
