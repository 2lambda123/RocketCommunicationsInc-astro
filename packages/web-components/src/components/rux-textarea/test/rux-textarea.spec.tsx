import { newSpecPage } from '@stencil/core/testing'
import { RuxTextarea } from '../rux-textarea'

describe('rux-textarea', () => {
    it('renders', async () => {
        const page = await newSpecPage({
            components: [RuxTextarea],
            html: `<rux-textarea></rux-textarea>`,
        })
        expect(page.root).toMatchSnapshot()
    })

    it('renders label prop', async () => {
        const page = await newSpecPage({
            components: [RuxTextarea],
            html: `<rux-textarea label="hello"></rux-textarea>`,
        })
        expect(page.root).toMatchSnapshot()
    })

    it('renders label slot', async () => {
        const page = await newSpecPage({
            components: [RuxTextarea],
            html: `<rux-textarea><div slot="label">hello</div></rux-textarea>`,
        })
        expect(page.root).toMatchSnapshot()
    })
})
