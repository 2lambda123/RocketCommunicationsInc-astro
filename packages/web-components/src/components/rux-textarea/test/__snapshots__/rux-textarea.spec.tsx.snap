// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rux-textarea renders 1`] = `
<rux-textarea value="">
  <mock:shadow-root>
    <div class="rux-textarea-field" part="form-field"><textarea aria-invalid="false" class="rux-textarea" id="rux-textarea-1" part="textarea" value=""></textarea>
    </div>
  </mock:shadow-root>
  <input class="aux-input" type="hidden" value="">
</rux-textarea>
`;

exports[`rux-textarea renders label prop 1`] = `
<rux-textarea label="hello" value="">
  <mock:shadow-root>
    <div class="rux-textarea-field" part="form-field">
      <label aria-hidden="false" class="rux-textarea-label" htmlfor="rux-textarea-2" part="label">
        <span>
          <slot name="label">
            hello
          </slot>
        </span>
      </label><textarea aria-invalid="false" class="rux-textarea" id="rux-textarea-2" part="textarea" value=""></textarea>
    </div>
  </mock:shadow-root>
  <input class="aux-input" type="hidden" value="">
</rux-textarea>
`;

exports[`rux-textarea renders label slot 1`] = `
<rux-textarea value="">
  <mock:shadow-root>
    <div class="rux-textarea-field" part="form-field">
      <label aria-hidden="false" class="rux-textarea-label" htmlfor="rux-textarea-3" part="label">
        <span>
          <slot name="label"></slot>
        </span>
      </label><textarea aria-invalid="false" class="rux-textarea" id="rux-textarea-3" part="textarea" value=""></textarea>
    </div>
  </mock:shadow-root>
  <div slot="label">
    hello
  </div>
  <input class="aux-input" type="hidden" value="">
</rux-textarea>
`;
